File: sculpt/prototype/sculpt/chisel.py
"""Chisel and NeighborStateKey (DESIGN.md §4.3, §4.4).

Chisel = absolute carving instruction at one level.
NeighborStateKey = hash key encoding self + 8 neighbor depths. Self-only
matching is forbidden (P3).
"""

from dataclasses import dataclass
from typing import List, Tuple

from .cell import Cell


def _quantize_self_channel(depth: int) -> int:
    # 4 bits per self channel = 16 buckets of 16.
    return (depth >> 4) & 0xF


def _quantize_neighbor(cell: Cell) -> int:
    # 3 bits per neighbor, averaged across 4 channels = 8 buckets of ~32.
    avg = (cell.depth_r + cell.depth_g + cell.depth_b + cell.depth_a) // 4
    return (avg >> 5) & 0x7


@dataclass(frozen=True)
class NeighborStateKey:
    self_r: int
    self_g: int
    self_b: int
    self_a: int
    n_tl: int
    n_t: int
    n_tr: int
    n_l: int
    n_r: int
    n_bl: int
    n_b: int
    n_br: int

    def pack_u64(self) -> int:
        # 4*4 + 8*3 = 40 bits, fits uint64 with room.
        v = 0
        for bits, val in (
            (4, self.self_r), (4, self.self_g), (4, self.self_b), (4, self.self_a),
            (3, self.n_tl), (3, self.n_t), (3, self.n_tr),
            (3, self.n_l),                   (3, self.n_r),
            (3, self.n_bl), (3, self.n_b), (3, self.n_br),
        ):
            v = (v << bits) | (val & ((1 << bits) - 1))
        return v


def build_neighbor_key(self_cell: Cell, neighbors_8: List[Cell]) -> NeighborStateKey:
    """Build the matching key. neighbors_8 is REQUIRED — signature enforces P3.

    Expected neighbor order (grid.DIRS_8):
      TL, T, TR, L, R, BL, B, BR
    """
    assert len(neighbors_8) == 8, "P3 violation: must supply all 8 neighbors"
    tl, t, tr, l, r, bl, b, br = neighbors_8
    return NeighborStateKey(
        self_r=_quantize_self_channel(self_cell.depth_r),
        self_g=_quantize_self_channel(self_cell.depth_g),
        self_b=_quantize_self_channel(self_cell.depth_b),
        self_a=_quantize_self_channel(self_cell.depth_a),
        n_tl=_quantize_neighbor(tl),
        n_t=_quantize_neighbor(t),
        n_tr=_quantize_neighbor(tr),
        n_l=_quantize_neighbor(l),
        n_r=_quantize_neighbor(r),
        n_bl=_quantize_neighbor(bl),
        n_b=_quantize_neighbor(b),
        n_br=_quantize_neighbor(br),
    )


@dataclass
class Chisel:
    chisel_id: int
    pre_state: NeighborStateKey
    subtract_r: int
    subtract_g: int
    subtract_b: int
    subtract_a: int
    level: int
    weight: int = 1
    usage_count: int = 0

    def subtract_tuple(self) -> Tuple[int, int, int, int]:
        return (self.subtract_r, self.subtract_g, self.subtract_b, self.subtract_a)

File: sculpt/src/chisel.c
#include "sculpt_chisel.h"
#include <assert.h>
#include <string.h>

static uint8_t quantize_self_channel(uint8_t depth)
{
    return (depth >> 4) & 0xF;  /* 4 bits = 16 buckets */
}

static uint8_t quantize_neighbor(const sculpt_cell_t *c)
{
    int avg = ((int)c->depth_r + c->depth_g + c->depth_b + c->depth_a) / 4;
    return (uint8_t)((avg >> 5) & 0x7);  /* 3 bits = 8 buckets */
}

void sculpt_neighbor_key_build(const sculpt_cell_t *self,
                                const sculpt_cell_t *neighbors[8],
                                sculpt_neighbor_key_t *out_key)
{
    /* P3: caller must supply non-null neighbors[0..7]. Assert to catch misuse. */
    assert(self != NULL);
    assert(neighbors != NULL);
    for (int i = 0; i < 8; ++i) {
        assert(neighbors[i] != NULL);
    }

    out_key->self_r = quantize_self_channel(self->depth_r);
    out_key->self_g = quantize_self_channel(self->depth_g);
    out_key->self_b = quantize_self_channel(self->depth_b);
    out_key->self_a = quantize_self_channel(self->depth_a);
    for (int i = 0; i < 8; ++i) {
        out_key->n[i] = quantize_neighbor(neighbors[i]);
    }
}

uint64_t sculpt_neighbor_key_pack(const sculpt_neighbor_key_t *k)
{
    uint64_t v = 0;
    v = (v << 4) | (k->self_r & 0xF);
    v = (v << 4) | (k->self_g & 0xF);
    v = (v << 4) | (k->self_b & 0xF);
    v = (v << 4) | (k->self_a & 0xF);
    for (int i = 0; i < 8; ++i) {
        v = (v << 3) | (k->n[i] & 0x7);
    }
    return v;  /* 4*4 + 8*3 = 40 bits */
}

int sculpt_neighbor_key_equal(const sculpt_neighbor_key_t *a,
                               const sculpt_neighbor_key_t *b)
{
    return memcmp(a, b, sizeof(*a)) == 0;
}
